### PR TITLE
Battery Checker Subsystem

### PR DESCRIPTION
**Pull request by calvinwt**
This will implement the battery subsystem (not simulation, because it will likely not be needed and will otherwise clutter up simulation logs)

Battery subsystem will monitor the battery and then send alerts if it dips below a certain voltage level or a certain amount of time passes

Develop was selected

### CODE REVIEW

**Review by azvanderpas — DISMISSED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -17,6 +17,7 @@
  * project.
  */
 public class Robot extends TimedRobot {
+  public static final boolean isCompetition = false;
```
[the remark this answers is not in the document]
Comment (caseyjbrotherton) [reply src/main/java/frc/robot/Robot.java:20]: Hmm.  Until our competition, we will be using main for driver practice.  
Do we think that we will need the battery subsystem and any other tuning to be available by default in driver practice?  
Or should driver practice be as close to competition experience as possible?

I think maybe the latter, so let's make the default that main always has isCOmpetition = true, and we change the default if we need more logging.

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -30,19 +31,25 @@ public class RobotContainer {
...
     swerveSubsystem.setDefaultCommand(new SwerveTeleopCommand(
       swerveSubsystem, 
       OI.Driver.getXTranslationSupplier(),
       OI.Driver.getYTranslationSupplier(),
       OI.Driver.getRotationSupplier()));
 
-      
+    //Implement when SmartDashboard/Whatever system we use is ready
+    //SmartDashboard.putData(batterySubsystem);
+
```
[the remark this answers is not in the document]
Comment (caseyjbrotherton) [reply src/main/java/frc/robot/RobotContainer.java:52]: +1 on card in trello, I would actually say that we can count on shuffleboard.  But smartdashboard shows up as a tab on shuffleboard...and we will never run battery checking in production.

In the card, note to make this controlled by isCompetition

File: src/main/java/frc/robot/Constants.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    // Battery Constants
+    public static final double minVoltageYellow = 11.0;
+    public static final double minVoltageRed = 12.0;
+    public static final double minVoltageRedDouble = 10.0;
+    public static final int timeInSecondsSimTest = 5;
+    public static final double timeInSecondsGeneralYellow = 360.0;
+    public static final double timeInSecondsGeneralRed = 600.0;
+    public static final double timeInSecondsHighCurrentYellow = 120.0;
+    public static final double timeInSecondsHighCurrentRed = 180.0;
+    public static final double highBatteryCurrentThreshold = 10;
```
[the remark this answers is not in the document]
Comment (calvinwt) [reply src/main/java/frc/robot/Constants.java:26]: finished, removed constants file and moved variables

File: src/main/java/frc/robot/Robot.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (calvinwt) [reply src/main/java/frc/robot/Robot.java:20]: finished, default value is now true

File: src/main/java/frc/robot/RobotContainer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (calvinwt) [reply src/main/java/frc/robot/RobotContainer.java:52]: finished, card added

File: src/main/java/frc/robot/subsystems/BatterySubsystem.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import edu.wpi.first.networktables.GenericEntry;
+import edu.wpi.first.networktables.NetworkTableEntry;
```
[the remark this answers is not in the document]
Comment (calvinwt) [reply src/main/java/frc/robot/subsystems/BatterySubsystem.java]: finished, removed

**Review by calvinwt — COMMENTED** [review]
All changes have been implemented

File: src/main/java/frc/robot/RobotContainer.java [hunk 5]
```diff
@@ -30,19 +31,25 @@ public class RobotContainer {
...
     swerveSubsystem.setDefaultCommand(new SwerveTeleopCommand(
       swerveSubsystem, 
       OI.Driver.getXTranslationSupplier(),
       OI.Driver.getYTranslationSupplier(),
       OI.Driver.getRotationSupplier()));
 
-      
+    //Implement when SmartDashboard/Whatever system we use is ready, make sure to bind it to isCompetition in Robot.java
+    //SmartDashboard.putData(batterySubsystem);
+
```
Comment (azvanderpas) [inline src/main/java/frc/robot/RobotContainer.java:52]: Is there a reason we can't put this in with the subsystem instantiation already?

**Comment by azvanderpas** [comment]
Missing simulation code

File: src/main/java/frc/robot/RobotContainer.java (hunk 5 above)
Comment (calvinwt) [reply src/main/java/frc/robot/RobotContainer.java]: done, just removed it because we already use shuffleboard

**Review by calvinwt — COMMENTED** [review]
Completed requested changes

**Review by caseyjbrotherton — COMMENTED** [review]
All changes completed, pushing

**Review by caseyjbrotherton — APPROVED** [review]

**Pull request merged**